quinn-proto/src/connection/assembler.rs

```diff
@@ -197,6 +239,50 @@ impl Chunk {
 struct Buffer {
     offset: u64,
     bytes: Bytes,
+    allocation_size: usize,
+    defragmented: bool,
+}
+
+impl Buffer {
+    /// Constructs a new fragmented Buffer
+    fn new(offset: u64, bytes: Bytes, allocation_size: usize) -> Self {
+        Self {
+            offset,
+            bytes,
+            allocation_size,
+            defragmented: false,
+        }
+    }
+
+    /// Constructs a new defragmented Buffer
+    fn new_defragmented(offset: u64, bytes: Bytes) -> Self {
+        let allocation_size = bytes.len();
+        Self {
+            offset,
+            bytes,
+            allocation_size,
+            defragmented: true,
+        }
+    }
+
+    /// Discards data before `offset` and flags `self` as defragmented if it has good utilization
+    fn try_mark_defragment(&mut self, offset: u64) {
+        let duplicate = offset.saturating_sub(self.offset) as usize;
+        self.offset = self.offset.max(offset);
+        if duplicate >= self.bytes.len() {
+            self.bytes = Bytes::new();
+            self.defragmented = true;
+            self.allocation_size = 0;
+            return;
+        } else {
+            self.bytes.advance(duplicate);
+        }
```

Comment:
No `else` needed anymore, nor do we still need the `self.defragmented ||` below.

Reply:
Got rid of the `else` branch. However, the `self.defragmented ||` is required to make sure that defragmented buffers remain defragmented. I added a comment stating that. If a defragmented buffer could become fragmented again, we might end up copying from a `Bytes` allocated by a previous `defragment()`, which does not get rid of the previous allocation because it might be shared between multiple buffers.

Reply:
Ah, you're right.